src/Rel8/Query/Evaluate.hs

```diff
@@ -26,57 +24,42 @@ import qualified Opaleye.Internal.Unpackspec as Opaleye
 
 -- rel8
 import Rel8.Expr ( Expr )
+import Rel8.Expr.Bool ( (&&.) )
+import Rel8.Expr.Opaleye ( fromPrimExpr )
 import Rel8.Query ( Query( Query ) )
 import Rel8.Table ( Table )
+import Rel8.Table.Bool ( case_ )
 import Rel8.Table.Opaleye ( unpackspec )
+import Rel8.Table.Undefined
 
--- semigroupoids
-import Data.Functor.Apply ( Apply )
-import Data.Functor.Bind ( Bind, (>>-) )
 
--- transformers
-import Control.Monad.Trans.State.Strict ( State, get, put, runState )
+-- | 'evaluate' takes expressions that could potentially have side effects and
+-- \"runs\" them in the 'Query' monad. The returned expressions have no side
+-- effects and can safely be reused.
+evaluate :: Table Expr a => a -> Query a
+evaluate = laterally >=> rebind
 
 
-type Evaluations :: Type
-data Evaluations = Evaluations
-  { tag :: !Opaleye.Tag
-  , bindings :: !(Endo (Opaleye.Bindings Opaleye.PrimExpr))
-  }
+laterally :: Table Expr a => a -> Query a
+laterally a = Query $ \bindings -> pure $ (Any True,) $
+  case nonEmpty bindings of
+    Nothing -> a
+    Just bindings' -> case_ [(condition, a)] undefined
+      where
+        condition = foldl1' (&&.) (fmap go bindings')
+          where
+            go = fromPrimExpr . Opaleye.UnExpr Opaleye.OpIsNotNull
 
 
--- | Some PostgreSQL functions, such as 'Rel8.nextval', have side effects,
--- breaking the referential transparency we would otherwise enjoy.
---
--- To try to recover our ability to reason about such expressions, 'Evaluate'
--- allows us to control the evaluation order of side-effects by sequencing
--- them monadically.
-type Evaluate :: Type -> Type
-newtype Evaluate a = Evaluate (State Evaluations a)
-  deriving newtype (Functor, Apply, Applicative, Monad)
-
-
-instance Bind Evaluate where
-  (>>-) = (>>=)
-
-
--- | 'eval' takes expressions that could potentially have side effects and
--- \"runs\" them in the 'Evaluate' monad. The returned expressions have no
--- side effetcs and can safely be reused.
-eval :: Table Expr a => a -> Evaluate a
-eval a = Evaluate $ do
-  Evaluations {tag, bindings} <- get
+rebind :: Table Expr a => a -> Query a
+rebind a = Query $ \_ -> Opaleye.QueryArr $ \(_, query, tag) ->
```

Comment:
I don't know what you mean. This is what makes it so that if you do `(\x -> (x, x)) <$> evaluate (nextval "user_id_seq")`, both your `x`s will have the same value. This is orthogonal to `laterally` which adds the superfluous lateral references.

Reply:
Added a comment anyway